File: zig-src/c/src/string.c
#include <stddef.h>

void *memcpy(void *restrict dest, const void *restrict src, size_t n) {
    if (!dest || !src || n == 0) return dest;

    unsigned char *d = (unsigned char *)dest;
    const unsigned char *s = (const unsigned char *)src;

    for (size_t i = 0; i < n; i++) {
        d[i] = s[i];
    }
    return dest;
}

void *memset(void *s, int c, size_t n) {
    if (!s || n == 0) return s;

    unsigned char *d = (unsigned char *)s;
    unsigned char value = (unsigned char)c;

    for (size_t i = 0; i < n; i++) {
        d[i] = value;
    }
    return s;
}

void *memmove(void *dest, const void *src, size_t n) {
    if (!dest || !src || n == 0) return dest;

    unsigned char *d = (unsigned char *)dest;
    const unsigned char *s = (const unsigned char *)src;

    if (d > s && d < s + n) {
        size_t i = n;
        while (i > 0) {
            i--;
            d[i] = s[i];
        }
    } else {
        for (size_t i = 0; i < n; i++) {
            d[i] = s[i];
        }
    }
    return dest;
}

int memcmp(const void *s1, const void *s2, size_t n) {
    if (!s1 || !s2 || n == 0) return 0;

    const unsigned char *p1 = (const unsigned char *)s1;
    const unsigned char *p2 = (const unsigned char *)s2;

    for (size_t i = 0; i < n; i++) {
        if (p1[i] != p2[i]) {
            return (p1[i] < p2[i]) ? -1 : 1;
        }
    }
    return 0;
}

size_t strlen(const char *s) {
    if (!s) return 0;

    size_t len = 0;
    while (s[len] != '\0') {
        len++;
    }
    return len;
}

int strcmp(const char *s1, const char *s2) {
    if (!s1 || !s2) return 0;

    size_t i = 0;
    while (s1[i] != '\0' && s1[i] == s2[i]) {
        i++;
    }

    return (unsigned char)s1[i] - (unsigned char)s2[i];
}

int strncmp(const char *s1, const char *s2, size_t n) {
    if (!s1 || !s2 || n == 0) return 0;

    for (size_t i = 0; i < n; i++) {
        if (s1[i] == '\0' || s1[i] != s2[i]) {
            return (unsigned char)s1[i] - (unsigned char)s2[i];
        }
    }
    return 0;
}

File: zig-src/string.zig
//! string.zig (适配现代 Zig 语法)

export fn memcpy(noalias dest: ?*anyopaque, noalias src: ?*const anyopaque, n: usize) callconv(.C) ?*anyopaque {
    if (dest == null or src == null or n == 0) return dest;

    const d: [*]u8 = @ptrCast(dest.?);
    const s: [*]const u8 = @ptrCast(src.?);

    var i: usize = 0;
    while (i < n) : (i += 1) {
        d[i] = s[i];
    }
    return dest;
}

export fn memset(s: ?*anyopaque, c: c_int, n: usize) callconv(.C) ?*anyopaque {
    if (s == null or n == 0) return s;

    const d: [*]u8 = @ptrCast(s.?);
    // 现代 Zig 自动推导 @intCast 目标类型
    const value: u8 = @intCast(c & 0xFF);

    var i: usize = 0;
    while (i < n) : (i += 1) {
        d[i] = value;
    }
    return s;
}

export fn memmove(dest: ?*anyopaque, src: ?*const anyopaque, n: usize) callconv(.C) ?*anyopaque {
    if (dest == null or src == null or n == 0) return dest;

    const d: [*]u8 = @ptrCast(dest.?);
    const s: [*]const u8 = @ptrCast(src.?);

    if (@intFromPtr(d) > @intFromPtr(s) and @intFromPtr(d) < @intFromPtr(s) + n) {
        var i: usize = n;
        while (i > 0) {
            i -= 1;
            d[i] = s[i];
        }
    } else {
        var i: usize = 0;
        while (i < n) : (i += 1) {
            d[i] = s[i];
        }
    }
    return dest;
}

export fn memcmp(s1: ?*const anyopaque, s2: ?*const anyopaque, n: usize) callconv(.C) c_int {
    if (s1 == null or s2 == null or n == 0) return 0;

    const p1: [*]const u8 = @ptrCast(s1.?);
    const p2: [*]const u8 = @ptrCast(s2.?);

    var i: usize = 0;
    while (i < n) : (i += 1) {
        if (p1[i] != p2[i]) {
            return if (p1[i] < p2[i]) -1 else 1;
        }
    }
    return 0;
}

export fn strlen(s: ?[*c]const u8) callconv(.C) usize {
    if (s == null) return 0;

    var len: usize = 0;
    while (s.?[len] != 0) : (len += 1) {}
    return len;
}

export fn strcmp(s1: ?[*c]const u8, s2: ?[*c]const u8) callconv(.C) c_int {
    if (s1 == null or s2 == null) return 0;

    var i: usize = 0;
    while (s1.?[i] != 0 and s1.?[i] == s2.?[i]) : (i += 1) {}

    const c1: c_int = s1.?[i];
    const c2: c_int = s2.?[i];
    return c1 - c2;
}

export fn strncmp(s1: ?[*c]const u8, s2: ?[*c]const u8, n: usize) callconv(.C) c_int {
    if (s1 == null or s2 == null or n == 0) return 0;

    var i: usize = 0;
    while (i < n) : (i += 1) {
        if (s1.?[i] == 0 or s1.?[i] != s2.?[i]) {
            const c1: c_int = s1.?[i];
            const c2: c_int = s2.?[i];
            return c1 - c2;
        }
    }
    return 0;
}
